source/loader/linux/loader_location.cpp

```diff
@@ -0,0 +1,81 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ */
+
+#include <dlfcn.h>
+#include <filesystem>
+#include <optional>
+
+#include "ur_loader.hpp"
+
+namespace fs = std::filesystem;
+
+namespace ur_loader {
+
+static std::optional<fs::path> getLoaderLibDirectory() {
+    Dl_info info;
+    if (dladdr((void *)getLoaderLibDirectory, &info)) {
+        auto libPath = fs::path(info.dli_fname);
+        if (fs::exists(libPath)) {
+            return fs::absolute(libPath).parent_path();
+        }
+    }
+
+    return std::nullopt;
+}
+
+static std::optional<std::vector<fs::path>> getEnvOSSearchPaths() {
+    std::optional<std::string> pathStringsOpt;
+    try {
+        pathStringsOpt = ur_getenv("LD_LIBRARY_PATH");
```

Comment:
I looked at the documentation for ld.so.8 (https://man7.org/linux/man-pages/man8/ld.so.8.html) and I'm actually not sure that interpreting LD_LIBRARY_PATH within UR is a good idea. There are a few special cases that we might not cover by interpreting it ourselves:
- "This variable is ignored in secure-execution mode"
- "Within the pathnames specified in LD_LIBRARY_PATH, the
              dynamic linker expands the tokens $ORIGIN, $LIB, and
              $PLATFORM (or the versions using curly braces around the
              names) as described above in Dynamic string tokens."
- The directory lists can also be separated by using semicolons
              
I think it would be safer to let the linker handle `LD_LIBRARY_PATH`.

Reply:
There might be something similiar on Windows as well.

Reply:
Since that was the case, I've changed the adapter registry to store a sorted list of paths that should be searched when looking for each hardcoded adapter. Eg. 1. Paths from `UR_ADAPTERS_SEARCH_PATH`. 2. Hardcoded library name (in that case the OS searches for this library). 3. UR Loader directory (current binary/library directory).